src/Simulation/qdk_sim_rs/docs/c-api.md

```diff
@@ -0,0 +1,147 @@
+# Using Experimental Simulators from C
+
+This module exposes a C API for this crate, useful for embedding into simulation
+runtimes.
+
+## Safety
+
+As this is a foreign-function interface, many of the functions exposed here
+are **unsafe**, representing that the caller is required to ensure that safety
+conditions in the host language are upheld.
+
+Please pay attention to any listed safety notes when calling into this C
+API.
+
+## Generating and Using C API Headers
+
+The [`qdk_sim`](..) crate has enabled the use of [`cbindgen`](https://crates.io/crates/cbindgen), such that C-language header files are generated automatically as part of the build for this crate.
+
+```bash
+cargo install --force cbindgen
+cbindgen --language C --output include/qdk_sim.h
+cbindgen --language C++ --output include/qdk_sim.hpp
+```
+
+This will generate `include/qdk_sim.h` and `include/qdk_sim.hpp, which can then be used from C and C++ callers, respectively. For example, to call from C:
+
+```c
+#include <stdio.h>
+#include "qdk_sim.h"
+
+int main() {
+    uintptr_t sim_id;
+    uintptr_t result0, result1;
+
+    init(2, "mixed", &sim_id);
+    h(sim_id, 0);
+    h(sim_id, 1);
+
+    m(sim_id, 0, &result0);
+    m(sim_id, 1, &result1);
+
+    printf("got %llu %llu", result0, result1);
+    destroy(sim_id);
+}
+```
+
+To build and run the above example using Clang on Windows:
+
+```bash
+$ clang example.c -Iinclude -Ltarget/debug -lqdk_sim -lws2_32 -lAdvapi32 -lUserenv
+got 1 1
+```
+
+## Error Handling and Return Values
+
+Most C API functions for this crate return an integer, with `0` indicating success and any other value indicating failure. In the case that a non-zero value is returned, API functions will also set the last error message, accessible by calling [`lasterr`]:
+
+```c
+#include <stdio.h>
+#include "qdk_sim.h"
+
+int main() {
+    uintptr_t sim_id;
+    uintptr_t result0, result1;
+
+    if (init(2, "invalid", &sim_id) != 0) {
+        printf("Got an error message: %s", lasterr());
+    }
+
+    destroy(sim_id);
+}
+```
```

Comment:
Nit: I assume that `init()` and `destroy()` are a pair of corresponding functions (the same way as `malloc()` and `free()`, or constructor and destructor). If the first function of the pair (`malloc()`, constructor) _fails_, then the second function of the pair (`free()`, destructor) _is not called_. 
In your example the `destroy()` is called unconditionally (even if `init()` fails).

Reply:
Fair, agreed. My point with this was to show a minimal example to demonstrate conventions, but agreed that the example could be improved. That said, calling `destroy` is safe even if `init` has not been called:

```rust
#[no_mangle]
pub extern "C" fn destroy(sim_id: usize) -> i64 {
    as_capi_err(|| {
        let state = &mut *STATE.lock().unwrap();
        if state.contains_key(&sim_id) {
            state.remove(&sim_id);
            Ok(())
        } else {
            Err(format!("No simulator with id {} exists.", sim_id))
        }
    })
}
```

If `init` has not been called for a given `sim_id`, then the `state.contains_key` check fails, such that an error code is returned to the C caller, and an actionable error message is set to `lasterr`. On the other hand, using an uninitialized `sim_id` on the C side could cause problems, such that I still agree with improving the example here.

Reply:
```suggestion
```c
#include <stdio.h>
#include "qdk_sim.h"

int main() {
    uintptr_t sim_id;
    uintptr_t result0, result1;

    if (init(2, "invalid", &sim_id) != 0) {
        printf("Got an error message: %s", lasterr());
    } else {
        destroy(sim_id);
    }
}
```
```

Reply:
Looks better.
Consider also:
```c
    if (init(2, "invalid", &amp;sim_id) != 0) {
        printf("Got an error message: %s", lasterr());
        return 1;
    }
    destroy(sim_id);
    return 0;
}
```
In the earlier examples the `main()` was not returning the value (I've just noticed).

My example may clutter the minimalistic illustration of the idea with the unrelated fragments. Feel free to ignore and stay minimalistic.